### PR TITLE
Add Advanced logger for Non DXE phases

### LINKED ISSUE #522: [Feature]: Enable Advanced Logger in PEI and MM on Q35 and SBSA
### Feature Overview

Advanced Logger is enabled in DXE on Q35 and SBSA, and it should be enabled in all phases.

### Solution Overview

Add PEIM, PEI_CORE, MM_CORE, and MM_STANDALONE instances of AdvancedLoggerLib and the Advanced Logger DebugLib.

### Alternatives Considered

_No response_

### Urgency

Low

### Are you going to implement the feature request?

I will implement the feature

### Do you need maintainer feedback?

No maintainer feedback needed

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by kuqin12**
## Description

This change introduces non DXE advanced logger, which should enable us to verify the advanced logger related changes better. Resolves #522.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

This change was tested on both QEMU Q35 and SBSA packages and verified bootable to UEFI shell.

## Integration Instructions

N/A

### CODE REVIEW

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.fdf [hunk 1]
```diff
@@ -573,6 +573,8 @@ FILE FREEFORM = PCD(gZeroTouchPkgTokenSpaceGuid.PcdZeroTouchCertificateFile) {
   INF MfciPkg/MfciDxe/MfciDxe.inf
   INF MsGraphicsPkg/OnScreenKeyboardDxe/OnScreenKeyboardDxe.inf
   INF MsGraphicsPkg/SimpleWindowManagerDxe/SimpleWindowManagerDxe.inf
+  INF AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumper.inf
+  INF MdeModulePkg/Universal/FvSimpleFileSystemDxe/FvSimpleFileSystemDxe.inf
```
[the remark this answers is not in the document]
Comment (kuqin12) [reply Platforms/QemuQ35Pkg/QemuQ35Pkg.fdf:577]:
```suggestion
```

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc [hunk 2]
```diff
@@ -1187,6 +1189,8 @@ QemuQ35Pkg/Library/ResetSystemLib/StandaloneMmResetSystemLib.inf
   MdeModulePkg/Universal/HiiDatabaseDxe/HiiDatabaseDxe.inf
   MdeModulePkg/Universal/SetupBrowserDxe/SetupBrowserDxe.inf
   MdeModulePkg/Universal/MemoryTest/NullMemoryTestDxe/NullMemoryTestDxe.inf
+  AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumper.inf
+  MdeModulePkg/Universal/FvSimpleFileSystemDxe/FvSimpleFileSystemDxe.inf
```
[the remark this answers is not in the document]
Comment (kuqin12) [reply Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc:1193]:
```suggestion
```

File: Platforms/QemuSbsaPkg/QemuSbsaPkg.dsc [hunk 3]
```diff
@@ -599,6 +610,7 @@
   gEfiMdePkgTokenSpaceGuid.PcdMaximumLinkedListLength|0
   gEfiMdePkgTokenSpaceGuid.PcdSpinLockTimeout|10000000
   gEfiMdePkgTokenSpaceGuid.PcdUefiLibMaxPrintBufferSize|320
+  gAdvLoggerPkgTokenSpaceGuid.PcdAdvancedLoggerPreMemPages|1
```
Comment (kuqin12) [inline Platforms/QemuSbsaPkg/QemuSbsaPkg.dsc:613]: Try increasing this.

**Comment by makubacki** [comment]
@kuqin12, what is the current state of this PR?

File: Platforms/QemuSbsaPkg/QemuSbsaPkg.dsc (hunk 3 above)
Comment (kuqin12) [reply Platforms/QemuSbsaPkg/QemuSbsaPkg.dsc]: Updated

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc [hunk 4]
```diff
@@ -682,6 +680,10 @@
   gAdvLoggerPkgTokenSpaceGuid.PcdAdvancedFileLoggerFlush|3
   gEfiSecurityPkgTokenSpaceGuid.PcdUserPhysicalPresence|FALSE
 
+  gAdvLoggerPkgTokenSpaceGuid.PcdAdvancedLoggerPreMemPages|3
+  gAdvLoggerPkgTokenSpaceGuid.PcdAdvancedLoggerMessageMmCorePrefix|"[MM]"
```
[the remark this answers is not in the document]
Comment (kuqin12) [reply Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc]: Removed.

**Comment by kuqin12** [comment]
> @kuqin12, what is the current state of this PR?

Should be ready to go. Tested on both Q35 and SBSA.

**Review by TaylorBeebe — APPROVED** [review]

**Comment by makubacki** [comment]
Regarding #522. It states:

> Advanced Logger is enabled in DXE on Q35 and SBSA, and it should be **enabled in all phases**.

> Add PEIM, PEI_CORE, **MM_CORE**, and **MM_STANDALONE** instances of AdvancedLoggerLib and the Advanced Logger DebugLib.

For future reference, I think the issue should mention that for QemuSbsaPkg, `MM_CORE_STANDALONE` and `MM_STANDALONE` are not covered due to support being limited to TFA.

**Review by makubacki — APPROVED** [review]

**Comment by kuqin12** [comment]
> Regarding #522. It states:
> 
> > Advanced Logger is enabled in DXE on Q35 and SBSA, and it should be **enabled in all phases**.
> 
> > Add PEIM, PEI_CORE, **MM_CORE**, and **MM_STANDALONE** instances of AdvancedLoggerLib and the Advanced Logger DebugLib.
> 
> For future reference, I think the issue should mention that for QemuSbsaPkg, `MM_CORE_STANDALONE` and `MM_STANDALONE` are not covered due to support being limited to TFA.

Ah, yes. I was wondering whether I should actually create a new one for that. But that will be a deeper conversation of how we want to pursue TFA modifications. I will at least add a comment there for better reference.

**Pull request merged**